go/vt/wrangler/reparent.go

```diff
@@ -518,17 +518,25 @@ func (wr *Wrangler) plannedReparentShardLocked(ctx context.Context, ev *events.R
 		// Promote the selected candidate to master.
 		promoteCtx, promoteCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
 		defer promoteCancel()
-		rp, err := wr.tmc.PromoteSlave(promoteCtx, masterElectTabletInfo.Tablet)
+		rp, err := wr.tmc.PromoteReplica(promoteCtx, masterElectTabletInfo.Tablet)
 		if err != nil {
 			return vterrors.Wrapf(err, "failed to promote %v to master", masterElectTabletAliasStr)
 		}
 		reparentJournalPos = rp
 	} else if topoproto.TabletAliasEqual(currentMaster.Alias, masterElectTabletAlias) {
-		refreshCtx, refreshCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
-		defer refreshCancel()
+		// It is possible that a previous attempt to reparent failed to SetReadWrite
+		// so call it here to make sure underlying mysql is ReadWrite
+		rwCtx, rwCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+		defer rwCancel()
 
+		if err := wr.tmc.SetReadWrite(rwCtx, masterElectTabletInfo.Tablet); err != nil {
```

Comment:
I can't put my finger on it, but this makes me feel vaguely uneasy. Is it clear in your mind how we're confident at this point that no other tablet's mysqld is read-write? I'm having trouble putting together that chain of reasoning.

Reply:
Let's replay the steps:
1. Mark old master read-only
2. Designate new master in topo (mastership+timestamp)
3. Mark new master read-write
4. Do the rest of the steps

If something fails after step 1, the topo will return the old one as master, and we just have to PRS on it.
If something fails after step 2, the topo will return the new one as master, and we perform steps 3 onwards.

This is probably your discomfort (which I also feel a bit): Is there a situation where the topo will return the old one as master after we complete step 2? By our reasoning of timestamps, it should not, because we compare all timestamps before identifying the true master.

I would feel slightly better if we marked the old master as replica before step 2, at least as best effort. But you and deepthi convinced me that it's unnecessary.

Reply:
This reasoning seems sound. I had included a new test that re-runs PRS on the candidate master after first simulating a failure in step 2. I could add more tests like that (simulating failures at other points in the sequence).

Reply:
I'm convinced as well. Thanks.